projects/npm-tools/packages/npm-scripts/src/sass/build.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+const fs = require('fs');
+const globby = require('globby');
+const r2 = require('liferay-r2');
+const path = require('path');
+const sass = require('sass');
+
+const log = require('../utils/log');
+
+const SASS_EXCLUDE = [
+	'**/_*.scss',
+	'**/build/**',
+	'**/classes/**',
+
+	// Copy from CSSBuilderArgs.java
+
+	'**/_diffs/**',
+	'**/.sass-cache*/**',
+	'**/.sass_cache_*/**',
+	'**/_sass_cache_*/**',
+	'**/_styled/**',
+	'**/_unstyled/**',
+	'**/css/aui/**',
+	'**/css/clay/**',
+	'**/tmp/**',
+];
+
+const getClayPaths = () => {
+	try {
+		return require('@clayui/css').includePaths;
+	} catch (e) {
+		return [];
+	}
+};
+
+const clayIncludePaths = getClayPaths();
+
+async function collectSassFiles(baseDir, excludes = []) {
+	const excludePaths = [
+		...SASS_EXCLUDE.map((exclude) => path.join(baseDir, exclude)),
+		...excludes.map((exclude) => path.join(baseDir, exclude)),
+	];
+
+	return await globby([
+		path.join(baseDir, '**/*.scss'),
+		...excludePaths.map((exclude) => `!${exclude}`),
+	]);
+}
+
+function buildSass(file, output) {
+	const newFileName = file
+		.split('/')
+		[file.split('/').length - 1].replace('.scss', '.css');
+
+	const outputFile = path.join(output, newFileName);
+
+	const result = sass.renderSync({
+		file,
+		includePaths: clayIncludePaths,
+		outFile: outputFile,
+		sourceMap: true,
+	});
+
+	return [outputFile, result.css, result.map];
+}
+
+function copyCss(filePath, newDirectory) {
+	if (!fs.existsSync(newDirectory)) {
+		fs.mkdirSync(newDirectory, {recursive: true});
+	}
+
+	fs.copyFileSync(filePath, path.join(newDirectory, path.basename(filePath)));
+}
+
+const SASS_DIR = '.sass-cache';
+
+async function main(baseDir, {excludes, outputDir = '', rtl = false}) {
+	const files = await collectSassFiles(baseDir, excludes);
+
+	if (!files.length) {
+		log(`BUILD CSS: No files found.`);
+
+		return;
+	}
+
+	for (let i = 0; i < files.length; i++) {
```

Comment:
@wincent do we allow this in portal code now? For some reason I remember a "liferay rule" where we didn't use `for of` or `for in`

Reply:
On 7.4 we no longer have to worry about it; it was only ever a problem on IE (and [only in specific edge cases](https://github.com/liferay/liferay-frontend-guidelines/issues/53#issuecomment-508029423)). In repos like this one you don't have to worry about it at all:

- The `no-for-of` rule is [still turned on in the preset](https://github.com/liferay/liferay-frontend-projects/blob/801917a87c8260a01c84ddb68b839f8c92ed04ac/projects/eslint-config/index.js#L92); but...
- We [have it turned off in this repo](https://github.com/liferay/liferay-frontend-projects/blob/801917a87c8260a01c84ddb68b839f8c92ed04ac/.eslintrc.js#L25).

At some point in the future we should just drop it from the preset. When 7.3 was released we laid the groundwork that would enable us to apply different Babel rules on different branches (ie. so that we can use the same version of `@liferay/npm-scripts` everywhere), so it would be a natural extension of that to allow the same kind of thing for ESLint (and even if we didn't want to do that, there's always `.eslintrc.js`).

Reply:
> On 7.4 we no longer have to worry about it
> We have it turned off in this repo.

![](https://media.giphy.com/media/l3q2zVr6cu95nF6O4/giphy.gif)